### ISSUE TITLE
[Request] Teal Support

### ISSUE THREAD

**Issue by qEagleStrikerp** [pkulchenko/ZeroBraneStudio#1152]
Teal is still relatively new, but it looks very promising. Currently there is no IDE for it with debugging support, and seeing as ZeroBraneStudio is most people's favorite Lua IDE, it would be nice to have Teal support for it (via Plugin maybe?).

**Comment by pkulchenko** [comment]
Looking at the [Moonscript plugin](https://github.com/pkulchenko/ZeroBranePackage/blob/master/moonscript.lua) may be a good first step. Assuming that lua modules (like Mobdebug and luasocket) can be loaded, then someone can start debugging from a Teal script to test this quickly (without any additional coding, just by following the instructions for [remote debugging](https://studio.zerobrane.com/doc-remote-debugging)) and if Teal integration with Lua provides a replacement for load/loadstring and some support for source maps, then other debugging features can be supported as well (similar to how it's done for moonscript).

**Comment by qEagleStrikerp** [comment]
Thanks for your answer. I did indeed get it to work by altering the Moonscript plugin to create a Teal interpreter. But all syntax highlighting is lost for .tl files, sadly. If I write a Teal script and save it as .lua file, it compiles and has syntax highlighting. Is there any way to load a .tl file and still have syntax highlighting?

**Comment by pkulchenko** [comment]
> I got that to work by adding editor.specmap.tl = 'lua' to user.lua. So it is actually possible to set up with minimal work. Thank you!

Yes, that should work, but it's also possible to add a more elaborate lexer for the actual teal code. You can start with lualibs/lexers/lua.lua and turn it into tl.lua lexer (and then register it from teal.lua plugin using `ide:AddLexer("lexlpeg.teal", [[text of teal lexer]])` command). You'll also need to add it to a spec and register this spec with `ide:AddSpec()`:

```lua
local spec = {
  exts = {"tl"},
  lexer = "lexlpeg.teal", -- same lexer name as used in AddLexer
  apitype = "lua",
  linecomment = "--",
  sep = ".:",
  marksymbols = ide.spec.lua.marksymbols,
}
ide:AddSpec("teal", spec)
```

**Comment by pkulchenko** [comment]
> Should I create a pull request for the teal.lua plugin?

Yes, you can submit it to https://github.com/pkulchenko/ZeroBranePackage repository.

**Comment by qEagleStrikerp** [comment]
What's the syntax of ide:AddLexer(name, lexer)? Specifically, what does it expect for "lexer", a path to a file?

**Comment by pkulchenko** [comment]
> What's the syntax of ide:AddLexer(name, lexer)? Specifically, what does it expect for "lexer", a path to a file?

(for others who may be looking at this) It's a string with the lexer code to be loaded from a plugin.

> It seems that the function "ReloadAPIs" doesn't exist currently.
> The line that throws the error is "ide:AddSpec("teal", spec)".

Let me take a quick look.

**Comment by pkulchenko** [comment]
> "Error while processing file: 'src/editor/package.lua:1206: attempt to call global 'ReloadAPIs' (a nil value)'."

@qEagleStrikerp, the function does exist for sure (it's in src/editor/autocomplete.lua file), but you're probably making AddSpec call too early, before the setup is completed. You need to call it from `onRegister` plugin callback. See [this section in the documentation](https://studio.zerobrane.com/doc-plugin#registering-a-specification) for an example.

**Comment by qEagleStrikerp** [comment]
@pkulchenko This was indeed the case. Calling it within onRegister works. However, now I have another issue: ide:AddLexer() doesn't actually do anything, I get the same error message 
"Can't load LexLPeg 'lexlpeg.teal' lexer: no file 'D:\Portable\Coding\ZeroBraneStudio-1.90\lualibs\lexers\teal.lua'"
whether it is in the code or not. From what I could find out by looking at src/editor/editor.lua, it seems that this should create a new file called "lexlpeg.teal" on first calling, but this creation process never happens. Did I understand something wrong? I stored the entire lexer file in a string and passed it as the second argument to ide:AddLexer().

**Comment by qEagleStrikerp** [comment]
@pkulchenko Could you take a look at my previous Edit 4?

**Comment by pkulchenko** [comment]
Thank you for the update; good to see you're making progress! Colors are a bit problematic, as they are remapped using `lexlpegmap` in editor.lua, which is not configurable at this moment. I have it on my todo list, but it's not a simple change, so requires a bit of work. You *can* update all the styles in your own spec, so it should be fixable, but I'd prefer to have a more convenient way to map all those styles and their colors.

**Comment by qEagleStrikerp** [comment]
@pkulchenko Thanks for the quick answer! Is changing colors and changing the underlining / crossing out all related to the same problem? Because changing colors is not nearly as important for the visuals as changing the way the grammar is processed. Especially the following problem:
```
local enum foo
local enum bar
```
becomes

local enum foo
local <s>enum</s> bar

Which really destroys the optics.

**Comment by pkulchenko** [comment]
This is controlled by `marksymbols` element in the spec; you're using the Lua one with the obvious issues. You can remove it completely, which will disable it or provide your own Teal-specific function.

**Comment by qEagleStrikerp** [comment]
@pkulchenko Thanks, works like a charm, looks really beautiful. I created a pull request for the plugin :).

**Comment by qEagleStrikerp** [comment]
@pkulchenko I noticed there was a little feature missing, namely the ability to save both .lua and .tl files. I found a workaround, but I think the way I solved it is not the intended way (every other thing I tried messed up either the lua or the teal lexer though). Would be nice if you could take a look at it some time:
https://github.com/pkulchenko/ZeroBranePackage/pull/95/commits/1263d000bcb091cd7cb5e6af52da7acd159abfb2

**Comment by julianthurner** [comment]
@pkulchenko Any chance you could take a look at the aforementioned pull request? Then I could link to it from the Teal Repo.

**Comment by aubguillemette** [comment]
Bump

**Comment by ildar** [comment]
I'm looking at https://github.com/pkulchenko/ZeroBranePackage/pull/95 and
see it doesn't add `tl check` support. Am I overlooking it?

**Comment by qEagleStrikerp** [comment]
> I'm looking at [pkulchenko/ZeroBranePackage#95](https://github.com/pkulchenko/ZeroBranePackage/pull/95) and
> see it doesn't add `tl check` support. Am I overlooking it?

@ildar Teal has to be added to your systems PATH variable manually. Then you can access it via shell. However, since there is no way to access the system shell within ZeroBrane Studio (as far as I am concerned, at least), you won't be able to use `tl check` without a separate shell.

**Comment by ildar** [comment]
AFAIK tl module had check function though i don't know how to use it

**Comment by julianthurner** [comment]
@ildar 

> AFAIK tl module had check function though i don't know how to use it

Maybe you are mixing something up:

When installing Teal (and adding it to your systems PATH variable), you get access to Teal via **Powershell**:
`tl check myTealFile.tl`
`tl gen myTealFile.tl`

However, the aforementioned pull request adds a ZeroBraneStudio **Plugin** that enables one to write Teal files within ZeroBraneStudio. The only thing it does is run `.tl` files using Teal instead of Lua, and adds support for interpreting the language / highlighting keywords with colors. Teal still has to be installed separately, the plugin only adds support for using it, not Teal itself.

So there's no option to use `tl check myTealFile.tl` from within ZeroBraneStudio as that is a command that is usually used within a console / Powershell, and ZeroBraneStudio has no console built-in.

**Comment by qEagleStrikerp** [comment]
@ildar 

> AFAIK tl module had check function though i don't know how to use it

Maybe you are mixing something up:

When installing Teal (and adding it to your system's PATH variable), you get access to Teal via **Powershell**:
`tl check myTealFile.tl`
`tl gen myTealFile.tl`

However, the aforementioned pull request adds a ZeroBraneStudio **Plugin** that enables one to write Teal files within ZeroBraneStudio. The only thing it does is run `.tl` files using Teal instead of Lua, and adds support for interpreting the language / highlighting keywords with colors. Teal still has to be installed separately, the plugin only adds support for using it, not Teal itself.

So there's no option to use `tl check myTealFile.tl` from within ZeroBraneStudio as that is a command that is usually used within a console / Powershell, and ZeroBraneStudio has no console built-in.

**Comment by ildar** [comment]
I mean this:

$  lua
Lua 5.4.4  Copyright (C) 1994-2022 Lua.org, PUC-Rio
> tl=require 'tl'
> type(tl.check)
function

**Comment by qEagleStrikerp** [comment]
> I mean this:
> 
> $  lua
> Lua 5.4.4  Copyright (C) 1994-2022 Lua.org, PUC-Rio
>  tl=require 'tl'
>  type(tl.check)
> function

@ildar Sorry for the late reply, exams had been taking up every free time I had.
I just took a look at it and this is possible by placing "teal.lua" in "ZeroBraneStudio/lua". I added instructions to the Teal package, accordingly.
